util/dvsim/Runner.py

```diff
@@ -0,0 +1,240 @@
+                continue
+
+            status_changed = True
+            if item.status == "D":
+                continue
+
+            if item.status != "P":
+                # Kill its sub items if item did not pass.
+                item.set_sub_status("K")
+                log.error("[%s]: [%s]: [status] [%s: %s]",
+                          hms, item.target, item.identifier, item.status)
+            else:
+                log.log(VERBOSE, "[%s]: [%s]: [status] [%s: %s]",
+                        hms, item.target, item.identifier, item.status)
+
+            # Queue items' sub-items if it is done.
+            for sub_item in item.sub:
+                self.add_item(sub_item)
+
+        return status_changed
+
+    def dispatch(self):
+        '''Dispatch some queued items if possible'''
+        num_slots = min(Runner.slot_limit,
+                        Runner.max_parallel - Deploy.dispatch_counter,
+                        len(self.queued_items))
+        if not num_slots:
+            return
+
+        items = self.queued_items[0:num_slots]
+        self.queued_items = self.queued_items[num_slots:]
+        self.dispatched_items.extend(items)
+
+        tgt_names = OrderedDict()
+        for item in items:
+            if item.status is None:
+                tgt_names.setdefault(item.target, []).append(item.identifier)
+                item.dispatch_cmd()
+
+        hms = self.timer.hms()
+        for target, names in tgt_names.items():
+            log.log(VERBOSE, "[%s]: [%s]: [dispatch]:\n%s",
+                    hms, target, ", ".join(names))
+
+    def check_if_done(self, print_status):
+        '''Update the "done" flag for each TargetStatus object
+
+        If print_status or we've reached a time interval then print current
+        status for those that weren't known to be finished already.
+
+        '''
+        if self.timer.check_time():
+            print_status = True
+
+        hms = self.timer.hms()
+
+        all_done = True
+        for target, tgt_status in self.status.items():
+            was_done = tgt_status.done
+            tgt_status.check_if_done()
+            is_done = tgt_status.done
+
+            all_done &= is_done
+
+            if print_status and not (was_done and is_done):
+                stats = tgt_status.counters
+                width = "0{}d".format(len(str(stats["T"])))
+                msg = "["
+                for s in stats.keys():
+                    msg += s + ": {:{}}, ".format(stats[s], width)
+                msg = msg[:-2] + "]"
+                log.info("[%s]: [%s]: %s", hms, target, msg)
+        return all_done
+
+    def run(self):
+        '''Run all items'''
+
+        # Print the legend just once (at the start of the first run)
+        if Runner.print_legend:
+            log.info("[legend]: [Q: queued, D: dispatched, "
+                     "P: passed, F: failed, K: killed, T: total]")
+            Runner.print_legend = False
```

Comment:
Previously, `Deploy.deploy()` was invoked multiple times (later for coverage). Hence the `print_legend` flag was added. With these updates, it is no longer needed.